akka-actor-typed-tests/src/test/java/akka/actor/typed/javadsl/ActorCompile.java

```diff
@@ -38,7 +38,7 @@ public MyMsgB(String greeting) {
   Behavior<MyMsg> actor5 = ignore();
   Behavior<MyMsg> actor6 = tap((ctx, signal) -> {}, (ctx, msg) -> {}, actor5);
   Behavior<MyMsgA> actor7 = actor6.narrow();
-  Behavior<MyMsg> actor8 = deferred(ctx -> {
+  Behavior<MyMsg> actor8 = onStart(ctx -> {
```

Comment:
as chatted back then in gitter, my worry about "on" prefix here is that it does not sound like the other behaviours, but it does sound like onMessage and onSignal, which makes it confusing.

Sticking to preStart would be nice to my ear... also easier to mental map from old actors, and for new people easy to find it since the word "start". and the purpose being mostly to "do stuff before the real behavior starts".

Reply:
I kind of bought Heikos argument against `preStart`, I also think that it expresses very clearly what it does - "on start, do this to produce the behavior", just as `onSignal` expresses "when there is a signal, do that"

Reply:
`Behaviors.factory(...)` ?

Reply:
reworded my comment above, not quite happy with some API decisions we have, but let's leave that for a different discussion. in this one I don't think introducing new words such as when or create are a good idea. 

create is simply the "because we're missing apply" and does not carry special meaning.
onStart conflates the callback naming with the factory naming which I don't like since it stops me from the mental shortcut that all the on* calls are handling messages. 
preStart is established for "do before starting", which is good for existing users. however when focusing on new users, they would find this anyway due to the start in the name. it also expressed that this is not THE behaviour but rather something that prepares one, which is how deferred is used. 
mixing context into the name could work, but only if we made it THE way to context, which it is not because we pass the context in all callbacks. so it is confusing to say "onContext" or whichever version of those naming, as it explains the api limitation, and not the intent why one would call this method -- which is to.prepare stuff.

anyway, that's my rationale for preStart. 
I can live with onStart I guess, due to.lack of better candidates, but I find it has noticeable downsides.

Reply:
I'm also relatively ok with `preStart` mostly for the reason that it already is a coined actor term (and the types tells us it is not only for side effects anyway) but I think `onStart` better describes what it does (for me "pre" means before something happens, and how can anything happen before we start the actor, and also when before the actor starts is it, right away, or a bit later?)

Reply:
> onStart conflates the callback naming with the factory naming which I don't like since it stops me from the mental shortcut that all the on* calls are handling messages.

In some way I agree. On the other hand, the other signals are also not dealing with messages but handling of life-cycle events.

As a though-experiment, could we introduce a `PreStart` signal and implement `deferred` as `onSignal { case (ctx, PreStart) => factory(ctx) }`? If yes, then `onStart` would just be a shortcut for that pattern and would work well. I wonder then, what "start" means exactly. If we look at all the places `undefer` is called, "start" would really mean "start of the behavior" and not "start of the actor" (as it is with existing `preStart`). Is that the semantics we are after?

Reply:
A `PreStart` signal would be about actor start and not behavior start though, so that doesn't match `deferred`

Reply:
Initially there was a PreStart signal and we decided to remove it because it was complicated (ensure it was always the first signal/message and things around that, I think it was).

It would also not be possible to close over the things (e.g. child actor ref) you create in the `deferred` block, which is pretty convenient. Passing around ctx is in general something I think we should discourage.

It is start of the Behavior, since it can also be used when returning from ordinary message sending. Especially when composing behaviors, e.g. a library might want to use deferred for that case. However, most common is to use it at the top, which then means start or restart of the actor.

Reply:
Then I think neither `preStart` nor `onStart` are good names because it's unclear whether the behavior or the actor is meant. Especially "starting a behavior" seems not like a useful concept as the behavior is only a description of what should be run in the actor.

Then how about `Behavior.setup`? ;)

Reply:
I like `setup`

Reply:
Not sure brings any more clarity than `deferred` though.

Reply:
I think it does. deferred is delayed in my book. setup is more like prepare.

random definition: http://www.future-perfect.co.uk/grammar-tip/is-it-setup-set-up-or-set-up/

Reply:
I still think that `deferred` is simply a factory to create a behavior. Therefore I suggest `apply` for Scala and `create` for Java. These are well established terms in the Java-Scala ecosystem.

`setup` is better than `preStart` or `onStart`, but still less intuitive than `apply`/`create`.

Reply:
I think `apply`/`create` are bad because they do not usually take a factory to create the instance, but rather create the instance right away.

With a nights sleep on it I like `setup` better (given that "setup is more like prepare", what about `prepare`?).

Reply:
I've argued against apply/create above specifically - it is not how we use those words, and it's not discoverable nor signals the intent of this method. It specifically is the "extra one" you need sometimes, but not always to create behaviors -- after all `.immutable` is a 'create' as well.

`setup` is good, pretty smart actually. Avoids the problem of thinking what's being started, I'd be ok with that. It relies the intent of this method as well

Reply:
@johanandren I agree. `apply`/`create` are no good here. `setup` is fine and also sounds familiar. `prepare` otoh is a little too exotic for me.

Reply:
Seems we have some form of consensus on `setup` then. Thanks Johannes!